firebase-storage/src/main/java/com/google/firebase/storage/network/NetworkRequest.java

```diff
@@ -116,18 +115,18 @@ public static String getDefaultURL(@NonNull Uri gsUri) {
         + "/b/"
         + gsUri.getAuthority()
         + "/o/"
-        + (pathWithoutBucket != null ? Slashes.unSlashize(pathWithoutBucket) : "");
+        + (pathWithoutBucket != null ? Uri.encode(pathWithoutBucket) : "");
```

Comment:
Rather than getDefaultURL() returning a string, could it instead return an android.net.Uri? This would allow you to:
* Be able to use Uri.Builder instead of string manipulation in order to construct the uri.
* Not have to worry about whether the returned string is an encoded or decoded string.

The same argument applies throughout this file. Basically, anywhere a uri is created via string manipulation, you could change it to use Uri.Builder. I think that would make it easier to track encoded/decoded status and ensure valid uris are being generated.

At some point, you'd still have to convert it to a java.net.URL though. But I think the conversion is just:
```
Uri uri = ...;
URL url = new URL(uri.toString());
```
(And I *think* that properly takes care of encoding urls? But definitely best to verify that.)

Reply:
I played around with your suggestions and I think I do like it better when Uri handles all encoding. It removes a lot of path and query parameter encoding and moves the responsibility to a class that likely has test coverage for all these things. I am a tiny bit afraid that this might introduce an obscure regression, but it is also not unlikely that the original version of this PR did. Overall, I removed a bunch of Uri to String to Uri conversion so everything should be safer now.